marlowe-playground-client/src/Blockly/Dom.purs

```diff
@@ -0,0 +1,253 @@
+  = Field String
+  -- A Value is visually represented as a label in the block that can be attached to a new block on the side.
+  -- For example "by", "the amount of" and "currency" in the "Deposit block"
+  | Value Block
+  -- A Statement is visually represented as the pluggable elements inside of a block. For example the contract inside
+  -- the root block or the Actions inside of a When block.
+  -- We represent the statement children as an Array of Blocks, but as you can see in the top level XML example,
+  -- each `<statement>` has a single block child, and siblings are represented by a `<next>` node nested inside.
+  -- In the initial version of this module, Next was a BlockChild constructor, making it easier to parse the
+  -- XML, but knowing that it would complicate the translation to Term, I decided to flatten the statement's childs
+  -- while parsing.
+  | Statement (Array Block)
+
+type ChildWithName
+  = Tuple String BlockChild
+
+data ReadDomError
+  = TypeMismatch Element String
+  | MissingProperty Element String
+  | SingleChildExpected Element Int
+  | RootElementNotFound String
+  | IncorrectSiblingNesting String
+
+-- NOTE: In some errors the element is not currently used to display the error. The idea is that we could later Change
+--       the signature to Effect String and traverse the parents of the element to provide error location information.
+explainError :: ReadDomError -> String
+explainError (TypeMismatch element expectedType) = "Element is of the wrong type (" <> show expectedType <> " expected, " <> show (Element.tagName element) <> " received)"
+
+explainError (MissingProperty element missingProperty) = "Element is missing required property " <> show missingProperty
+
+explainError (SingleChildExpected element elementCount) = "Element was expected to have a single child, and it had " <> show elementCount
+
+explainError (RootElementNotFound rootBlockName) = "The element with id " <> show rootBlockName <> " was not found."
+
+explainError (IncorrectSiblingNesting node) = "Incorrect <next> element found outside the scope of a <statement> and inside of a " <> node <> " node"
+
+-- | Read and parse the DOM nodes of a blockly workspace.
+getDom :: BlocklyState -> Effect (Either ReadDomError Block)
+getDom { blockly, workspace, rootBlockName } =
+  runExceptT do
+    rootElement <- liftEffect $ workspaceToDom blockly workspace
+    if Element.tagName rootElement /= "xml" then
+      throwError $ TypeMismatch rootElement "xml"
+    else do
+      -- The workspace can have many elements at the top level, we parse them all
+      -- but only return the one that has the same id as rootBlockName
+      childrens <- getChildrens rootElement
+      blocks <- traverse readAsBlock childrens
+      case find (eq rootBlockName <<< view (_1 <<< _id)) blocks of
+        Nothing -> throwError $ RootElementNotFound rootBlockName
+        Just (Tuple block nexts) ->
```

Comment:
you can pattern match on `[]` here I think

Reply:
done